R/slide.R

```diff
@@ -209,128 +219,213 @@ epi_slide <- function(
+        return(.data_group %>% filter(min(time_value) + before <= time_value & time_value <= max(time_value) - after))
+      } else {
+        return(.data_group)
+      }
+    }
+    return(.data_group %>% filter(FALSE))
+  }
 
-    slide_values <- vctrs::list_unchop(slide_values_list)
+  # Get stateful function that tracks ref_time_value per group and sends it to
+  # `f` when called.
+  f <- f_factory(available_ref_time_values)
 
+  if (time_type == "yearmonth" && identical(before, Inf)) {
+    # <yearmonth> - Inf is NA(s) rather than -Inf as a yearmonth; feed in -Inf manually
+    # (it will successfully be cast to -Inf as a yearmonth)
+    starts <- rep(-Inf, length(available_ref_time_values))
+    stops <- available_ref_time_values + after
+  } else {
+    starts <- available_ref_time_values - before
+    stops <- available_ref_time_values + after
+  }
+
+  # Compute the slide values. slider::hop_index will return a list of f outputs
+  # e.g. list(f(.slide_group_1, .group_key, .ref_time_value_1),
+  # f(.slide_group_1, .group_key, .ref_time_value_2), ...)
+  slide_values_list <- slider::hop_index(
+    .x = .data_group,
+    .i = .data_group$time_value,
+    .starts = starts,
+    .stops = stops,
+    .f = f,
+    .group_key, ...
+  )
 
-    if (
-      all(purrr::map_int(slide_values_list, vctrs::vec_size) == 1L) &&
-        length(slide_values_list) != 0L
-    ) {
-      # Recycle to make size stable (one slide value per ref time value).
-      # (Length-0 case also could be handled here, but causes difficulties;
-      # leave it to the next branch, where it also belongs.)
-      slide_values <- vctrs::vec_rep_each(slide_values, times = counts)
+  # Validate returned values.
+  return_types <- purrr::map_chr(slide_values_list, function(x) {
+    if (is.data.frame(x)) {
+      return("data.frame")
+    } else if (vctrs::obj_is_vector(x) && is.null(vctrs::vec_names(x))) {
+      return("vector")
     } else {
-      # (Loose) check on number of rows:
-      if (vctrs::vec_size(slide_values) != num_ref_rows) {
-        cli_abort(
-          "The slide computations must either (a) output a single element/row each, or
-          (b) one element/row per appearance of the reference time value in the local window."
-        )
-      }
+      return("other")
+    }
+  }) %>% unique()
+  # Returned values must be data.frame or vector.
+  if ("other" %in% return_types) {
+    cli_abort(
+      "epi_slide: slide computations must always return either data frames without rownames
+      or unnamed vectors (as determined by the vctrs package).",
+      class = "epiprocess__invalid_slide_comp_value"
+    )
+  }
+  # Returned values must all be the same type.
+  if (length(return_types) != 1L) {
+    cli_abort(
+      "epi_slide: slide computations must always return either a data.frame or a vector (as determined by the
+      vctrs package), but not a mix of the two.",
+      class = "epiprocess__invalid_slide_comp_value"
+    )
+  }
```

Comment:
I think we can potentially avoid the size checks on elements of `slide_values_list` + the `list_unchop` by using `slider::hop_index_vec` instead of `slider::hop_index`; `slider::hop_index_vec` expects size-1 vectors in the vctrs sense, and will combine them for us (hopefully more efficiently than we can with a `list_unchop` after the fact).

Reply:
I think I may just push a commit for this because it may simplify #509 implementation.

Reply:
Reconsidering here.  I guess `hop_index_vec` is basically doing `hop_index` +  unnaming + unchopping; no fancy efficiency stuff.  We can get better validation by name-checking + unchopping, or name-checking + other-checks + unchopping, or name-checking + unchopping&wrapping-error-messages-from-unchopping.